src/value/types.rs

```diff
@@ -0,0 +1,126 @@
+use std::fmt::Debug;
+use std::rc::Rc;
+
+use crate::ColumnType;
+
+enum Type {
+    Bool,
+    TinyInt,
+    SmallInt,
+    Int,
+    BigInt,
+    TinyUnsigned,
+    SmallUnsigned,
+    Unsigned,
+    BigUnsigned,
+    Float,
+    Double,
+    String,
+    Char,
+    Array(Box<Type>),
+}
+
+pub trait ValueTrait: Debug {
+    fn to_sql_string(&self) -> String;
+}
+
+pub trait ValueType: Sized {
+    fn type_name() -> String;
+    fn column_type() -> ColumnType;
+}
+
+#[derive(Debug, Clone)]
+pub struct Value(Rc<Box<dyn ValueTrait>>);
```

Comment:
Hey @ikrivosheev, thanks for the draft!!

One question come up in my mind. For `sea-query-binder` and `sea-query-driver`, how do we bind the value on db driver? Because we can't figure out what's the type that implemented `ValueTrait` without the help of [`downcast`](https://doc.rust-lang.org/std/any/trait.Any.html#method.downcast) and `Any` trait.

Reply:
@tyt2y3 ideas?

Reply:
@billy1624 this is the biggest problem... Another idea is move `sqlx` and `driver` into sea-query

Reply:
@Sytten ideas? 

@billy1624 I can make two PR with with `downcast` and with move `sqlx` and `driver` into sea-query.

Reply:
The chain of pings loll

Reply:
I wonder what's the performance implication of using `downcast`?

Reply:
This what I discussed on discord: https://discord.com/channels/873880840487206962/900758332187484160/1007648248745447524

Reply:
@tyt2y3 I think, I undestand!

Reply:
> keeping the enum when you have the boxed trait

I wanted to put primitives (and to some extend anything that is less than 128 bits and impl Copy) on the stack.

Reply:
I imagine it's like

```rs
// Inside the root of sea-query

trait ValueTrait {}

enum SeaValue {
    I32(i32),
    Object(Box<dyn ValueTrait>),
}

impl ValueTrait for SeaValue {}

// Inside sea-orm-binder

struct SqlxMySqlValue<T>(T)
where
    T: ValueTrait + for<'q> Encode<'q, MySql>;

impl<'p, T> Encode<'p, MySql> for SqlxMySqlValue<T>
where
    T: ValueTrait + for<'q> Encode<'q, MySql>,
{
    fn encode_by_ref(
        &self,
        buf: &mut <MySql as sqlx::database::HasArguments<'_>>::ArgumentBuffer,
    ) -> sqlx::encode::IsNull {
        self.0.encode_by_ref(buf)
    }
}
```

Reply:
And, maybe...

```rs
trait ValueTrait {}

enum SeaValue {
    I32(i32),
    Object(Box<dyn ValueTrait>),
}

impl ValueTrait for SeaValue {}

impl<'p> Encode<'p, MySql> for dyn ValueTrait
where
    Self: for<'q> Encode<'q, MySql>,
{
    fn encode_by_ref(
        &self,
        buf: &mut <MySql as sqlx::database::HasArguments<'p>>::ArgumentBuffer,
    ) -> sqlx::encode::IsNull {
        self.encode_by_ref(buf)
    }
}
```

Reply:
@billy1624 it seems like it won't work...

```
// Inside the root of sea-query

trait ValueTrait {}

enum Value {
    I32(i32),
    Object(Box<dyn ValueTrait>),
}

impl ValueTrait for Value {}

// Inside sea-query-rusqlite

#[derive(Clone, Debug)]
struct RusqliteValue<T>(T)
where
    T: ValueTrait + ToSql;

impl<T> ToSql for RusqliteValue<T>
where
    T: ValueTrait + ToSql,
{
    fn to_sql(&self) -> Result<ToSqlOutput<'_>> {
        self.0.to_sql()
    }
}

#[derive(Clone, Debug)]
pub struct RusqliteValues<T>(pub Vec<RusqliteValue<T>>);  // <--- I cannot put here enum Value because it does not impl ToSql and I cannot make this parameter generic and I cannot impl ToSql for Value because it from other trait
```